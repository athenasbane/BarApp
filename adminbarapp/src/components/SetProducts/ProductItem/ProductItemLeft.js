import React from 'react';
import {
    Grid,
    Typography,
    Button,
    TextField,
    FormControlLabel,
    Switch,
    makeStyles
} from '@material-ui/core';

const useStyles = makeStyles({
    root: {
        marginTop: '10px',
        marginBottom: '10px',
        width: '100%',
        textAlign: 'center'
    },
    titleBox: {
        marginBottom: '10px'
    },
    bordered: {
        border: '1px solid black'
    },
    margins: {
        margin: '15px',
        textAlign: 'center',
        justifyContent: 'center'
    }
});

const ProductItemLeft = (props) => {
    
    const classes = useStyles();
    return (
        <React.Fragment>
            <Grid container direction="row">
                <Grid 
                    container 
                    item 
                    justify="center" 
                    justifycontent="center" 
                    alignItems="center" 
                    direction="column" 
                    xs={12}>
                    <Grid item>
                        <Typography>Title:</Typography> 
                        <TextField
                        className={classes.root}
                        required
                        name="title"
                        label="Required"
                        variant="filled"
                        value={props.product.title}
                        onChange={event => props.handleChange({name: event.target.name, 
                            value: event.target.value, id: props.product._id})}
                        />
                    </Grid>
                    <Grid item>
                        <Typography>Category:</Typography> 
                        <TextField
                            className={classes.root}
                            required
                            name="category"
                            label="Required"
                            variant="filled"
                            value={props.product.category}
                            onChange={event => props.handleChange({name: event.target.name, 
                                value: event.target.value, id: props.product._id})}
                            />
                    </Grid>
                    <Grid item>
                        <Typography>Sub-Category:</Typography> 
                            <TextField
                                className={classes.root}
                                required
                                name="subCategory"
                                label="Required"
                                variant="filled"
                                value={props.product.subCategory}
                                onChange={event => props.handleChange({name: event.target.name, 
                                    value: event.target.value, id: props.product._id})}
                                />
                    </Grid>
                    <Grid className={classes.margins} item >
                        <FormControlLabel
                        control={
                        <Switch
                            checked={props.product.active}
                            onChange={event => props.handleChange({name: event.target.name, 
                                value: event.target.checked, id: props.product._id })}
                            name="active"
                            color="primary"
                        />
                        }label="Active" />
                    </Grid>
                    <Grid className={classes.margins} container item >
                        <Button onClick={() => props.removeProduct(props.product._id)} 
                        variant="contained">Delete</Button>
                    </Grid>
                    <Grid className={classes.margins} container item >
                        <Button onClick={() => props.saveProduct(props.product)} 
                        color="secondary" variant="contained">Save</Button>
                    </Grid>
                </Grid>
            </Grid>
        </React.Fragment>
    );
};

export default ProductItemLeft;