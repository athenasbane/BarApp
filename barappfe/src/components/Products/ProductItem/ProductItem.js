import React from 'react';
import PropTypes from 'prop-types';
import {
  Accordion,
  makeStyles,
  Typography,
  AccordionSummary,
  AccordionDetails,
} from '@material-ui/core';
import AddIcon from '@material-ui/icons/Add';
import ProductForm from './ProductForm/ProductForm';

const useStyles = makeStyles({
  root: {
    width: '100%',
    marginTop: '2px',
  },
  paperEven: {
    backgroundColor: '#C0C0C0',
    textAlign: 'center',
  },
  paperOdd: {
    backgroundColor: '#D3D3D3',
    textAlign: 'center',
  },
  subHeading: {
    marginLeft: '100px',
  },
});

export const ProductItem = ({ title, index, active, id }) => {
  const classes = useStyles();
  const [expanded, setExpanded] = React.useState(false);

  const handleChange = (title) => (event, isExpanded) => {
    setExpanded(isExpanded ? title : false);
  };

  return (
    <Accordion
      expanded={expanded === title}
      className={index % 2 === 1 ? classes.paperOdd : classes.paperEven}
      onChange={handleChange(title)}
      disabled={!active}
    >
      <AccordionSummary
        expandIcon={<AddIcon />}
        aria-controls={`${title}-content`}
        id={`${title}-header`}
      >
        <Typography>{title}</Typography>
        {!active ? <Typography className={classes.subHeading}>Sold Out</Typography> : null}
      </AccordionSummary>
      <AccordionDetails>
        <ProductForm title={title} id={id} handleChange={handleChange(title)} />
      </AccordionDetails>
    </Accordion>
  );
};

ProductItem.propTypes = {
  title: PropTypes.string.isRequired,
  index: PropTypes.number.isRequired,
  active: PropTypes.bool.isRequired,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

export default ProductItem;
