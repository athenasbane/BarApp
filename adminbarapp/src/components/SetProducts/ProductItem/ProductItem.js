import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Paper, Typography, Button, makeStyles } from '@material-ui/core';
import { connect } from 'react-redux';
import ProductItemOption from './ProductItemOption/ProductItemOption';
import ProductItemLeft from './ProductItemLeft';
import { loadOptions, saveNewOption } from '../../../Redux/thunks/option.thunk';
import ProductItemNewOption from './ProductItemOption/ProductItemNewOption';

const useStyles = makeStyles({
  bordered: {
    border: '1px solid black',
  },
});

const ProductItem = ({
  id,
  getOptions,
  saveNewOption,
  options,
  product,
  saveProduct,
  removeProduct,
  handleChange,
  productIndex,
}) => {
  const classes = useStyles();
  const [optionOpen, setOptionOpen] = React.useState(false);
  const [newOptionData, setNewOptionData] = React.useState({
    optionTitle: '',
    type: 'increment',
    minVol: 0,
    price: 1,
    selector: [],
    optionActive: false,
  });

  React.useEffect(() => {
    if (id) {
      getOptions(id);
    }
  }, [getOptions, id]);

  const handleChangeItem = (event) => {
    event.persist();
    if (event.target.name === 'selector') {
      console.log('here');
      return setNewOptionData((prevState) => ({
        ...prevState,
        [event.target.name]: event.target.value.split(', '),
      }));
    }
    if (event.target.name === 'optionActive') {
      console.log(event.target.checked);
      return setNewOptionData((prevState) => ({
        ...prevState,
        optionActive: event.target.checked,
      }));
    }
    setNewOptionData((prevState) => ({
      ...prevState,
      [event.target.name]: event.target.value,
    }));
  };

  const handleNewOptionSave = (id, option) => {
    saveNewOption(id, option);
    setNewOptionData({
      optionTitle: '',
      type: 'increment',
      minVol: 0,
      price: 1,
      selector: [],
      optionActive: false,
    });
    setOptionOpen(!optionOpen);
  };

  const optionsFilter = options.filter((el) => el.product === id);

  const optionsList = optionsFilter.map((option, index) => (
    <ProductItemOption key={option._id} index={index} option={option} handleChange={handleChange} />
  ));
  return (
    <Paper elevation={3} variant="outlined" square>
      <Grid justify="center" container item xs={12}>
        <Typography>{product.title}</Typography>
      </Grid>
      <Grid container direction="row">
        <Grid item xs={4}>
          <ProductItemLeft
            saveProduct={saveProduct}
            removeProduct={removeProduct}
            handleChange={handleChange}
            product={product}
            productIndex={productIndex}
          />
        </Grid>
        <Grid container item direction="column" xs={8}>
          <Grid className={classes.bordered} item>
            {optionsList}
          </Grid>
          <Grid item>
            <Button
              disabled={!product._id}
              onClick={() => setOptionOpen(!optionOpen)}
              variant="contained"
            >
              Add New Option
            </Button>
            {optionOpen ? (
              <ProductItemNewOption
                option={newOptionData}
                changeHandler={handleChangeItem}
                handleCancel={() => setOptionOpen(!optionOpen)}
                handleSave={() => handleNewOptionSave(product._id, newOptionData)}
              />
            ) : null}
          </Grid>
        </Grid>
      </Grid>
    </Paper>
  );
};

ProductItem.propTypes = {
  id: PropTypes.string.isRequired,
  getOptions: PropTypes.func.isRequired,
  saveNewOption: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.object),
  product: PropTypes.object.isRequired,
  saveProduct: PropTypes.func.isRequired,
  removeProduct: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  productIndex: PropTypes.number.isRequired,
};

const mapStateToProps = (state) => ({
  options: state.option.optionData,
});

const mapDispatchToProps = (dispatch) => ({
  getOptions: (id) => dispatch(loadOptions(id)),
  saveNewOption: (id, option) => dispatch(saveNewOption(id, option)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ProductItem);
