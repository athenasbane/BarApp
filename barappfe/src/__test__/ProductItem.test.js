import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { ProductItem } from '../components/Products/ProductItem/ProductItem';

const mockStore = configureStore();
const store = mockStore();

describe('Menu Tests', () => {
    test('renders correctly', () => {
        const wrapper = shallow(
            <Provider store={store}>
                <ProductItem title="test title" index={1} active={true}/>
            </Provider>
        );
        expect(wrapper).toMatchSnapshot();
    });
}); 