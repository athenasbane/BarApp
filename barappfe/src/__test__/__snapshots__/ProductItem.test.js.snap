// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Tests renders correctly 1`] = `ShallowWrapper {}`;
